File: src/infrastructure/timezonesupport/MarketHours.cs
using core.Shared.Adapters.Brokerage;
using TimeZoneConverter;

namespace timezonesupport
{
    public class MarketHoursAlwaysOn : IMarketHours
    {
        private static readonly IMarketHours _marketHours = new MarketHours();

        DateTimeOffset IMarketHours.GetMarketEndOfDayTimeInUtc(DateTimeOffset when)
            => _marketHours.GetMarketEndOfDayTimeInUtc(when);

        DateTimeOffset IMarketHours.GetMarketStartOfDayTimeInUtc(DateTimeOffset when)
            => _marketHours.GetMarketStartOfDayTimeInUtc(when);

        bool IMarketHours.IsMarketOpen(DateTimeOffset time) => true;

        DateTimeOffset IMarketHours.ToMarketTime(DateTimeOffset utc)
            => _marketHours.ToMarketTime(utc);

        DateTimeOffset IMarketHours.ToUniversalTime(DateTimeOffset eastern)
            => _marketHours.ToUniversalTime(eastern);
    }

    public class MarketHours : IMarketHours
    {
        private static TimeZoneInfo _easternZoneId = TZConvert.GetTimeZoneInfo("Eastern Standard Time");
        private static TimeSpan StartTime = new TimeSpan(9, 30, 0);
        private static TimeSpan EndTime = new TimeSpan(16, 0, 0);

        private static DateTimeOffset ConvertToEastern(DateTimeOffset when)
        {
            return TimeZoneInfo.ConvertTimeFromUtc(
                when.DateTime,
                _easternZoneId
            );
        }

        bool IMarketHours.IsMarketOpen(DateTimeOffset utc)
        {
            // 930-1600
            var eastern = TimeZoneInfo.ConvertTimeFromUtc(
                utc.DateTime,
                _easternZoneId
            );

            if (eastern.DayOfWeek == DayOfWeek.Saturday || eastern.DayOfWeek == DayOfWeek.Sunday)
            {
                return false;
            }

            var timeOfDay = eastern.TimeOfDay;

            return timeOfDay >= StartTime && timeOfDay <= EndTime;
        }

        DateTimeOffset IMarketHours.ToMarketTime(DateTimeOffset utc) => ConvertToEastern(utc);

        DateTimeOffset IMarketHours.ToUniversalTime(DateTimeOffset eastern) =>
            TimeZoneInfo.ConvertTimeToUtc(eastern.DateTime, _easternZoneId);

        DateTimeOffset IMarketHours.GetMarketEndOfDayTimeInUtc(DateTimeOffset eastern) =>
            TimeZoneInfo.ConvertTimeToUtc(
                eastern.Date.Add(EndTime),
                _easternZoneId
            );

        DateTimeOffset IMarketHours.GetMarketStartOfDayTimeInUtc(DateTimeOffset eastern) =>
            TimeZoneInfo.ConvertTimeToUtc(
                eastern.Date.Add(StartTime),
                _easternZoneId
            );
    }
}

File: src/infrastructure/timezonesupport.fs/MarketHours.fs
namespace timezonesupport

open System
open TimeZoneConverter
open core.fs.Adapters.Brokerage

type MarketHours() =
    
    let easternZoneId = TZConvert.GetTimeZoneInfo("Eastern Standard Time")
    let startTime = TimeSpan(9, 30, 0)
    let endTime = TimeSpan(16, 0, 0)
    
    let convertToEastern (time: DateTimeOffset) =
        TimeZoneInfo.ConvertTimeFromUtc(time.DateTime, easternZoneId)
    
    member _.IsMarketOpen(utc: DateTimeOffset) : bool =
        let eastern = TimeZoneInfo.ConvertTimeFromUtc(utc.DateTime, easternZoneId)
        
        if eastern.DayOfWeek = DayOfWeek.Saturday || eastern.DayOfWeek = DayOfWeek.Sunday then
            false
        else
            let timeOfDay = eastern.TimeOfDay
            timeOfDay >= startTime && timeOfDay <= endTime
    
    member _.ToMarketTime(utc: DateTimeOffset) : DateTimeOffset =
        DateTimeOffset(convertToEastern utc)
    
    member _.ToUniversalTime(eastern: DateTimeOffset) : DateTimeOffset =
        DateTimeOffset(TimeZoneInfo.ConvertTimeToUtc(eastern.DateTime, easternZoneId))
    
    member _.GetMarketEndOfDayTimeInUtc(eastern: DateTimeOffset) : DateTimeOffset =
        DateTimeOffset(TimeZoneInfo.ConvertTimeToUtc(eastern.Date.Add(endTime), easternZoneId))
    
    member _.GetMarketStartOfDayTimeInUtc(eastern: DateTimeOffset) : DateTimeOffset =
        DateTimeOffset(TimeZoneInfo.ConvertTimeToUtc(eastern.Date.Add(startTime), easternZoneId))
    
    interface IMarketHours with
        member this.IsMarketOpen time = this.IsMarketOpen(time)
        member this.ToMarketTime utc = this.ToMarketTime(utc)
        member this.ToUniversalTime eastern = this.ToUniversalTime(eastern)
        member this.GetMarketEndOfDayTimeInUtc eastern = this.GetMarketEndOfDayTimeInUtc(eastern)
        member this.GetMarketStartOfDayTimeInUtc eastern = this.GetMarketStartOfDayTimeInUtc(eastern)

type MarketHoursAlwaysOn() =
    
    let marketHours = MarketHours()
    
    interface IMarketHours with
        member _.GetMarketEndOfDayTimeInUtc time =
            (marketHours :> IMarketHours).GetMarketEndOfDayTimeInUtc(time)
        
        member _.GetMarketStartOfDayTimeInUtc time =
            (marketHours :> IMarketHours).GetMarketStartOfDayTimeInUtc(time)
        
        member _.IsMarketOpen _ = true
        
        member _.ToMarketTime utc =
            (marketHours :> IMarketHours).ToMarketTime(utc)
        
        member _.ToUniversalTime eastern =
            (marketHours :> IMarketHours).ToUniversalTime(eastern)
